### ISSUE TITLE
As a package author, I want to test that a specific code does not compile

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/sdk#41906]
In some situations, a package may do a specific task to ensure that something compile or not.

As all behaviors, such thing would benefit from being tested, to avoid regressions.

Testing that a code compiles is straightforward: Just write the code. If it compiles, the analyzer shouldn't complaIn.\
The difficult part is testing that a code does **not** compile.

Currently, we need to write the desired code inside a String, analyze it with `analyzer`, and parse the analysis result to extract the error report.

This is very tedious, slow, brittle, and hard to debug. This usually requires more effort than the benefits it brings.\
We are loosing syntax highlighting, auto-completion, and the error is unintelligible.\
It would be great if we could simplify that use-case.

____

One possibility could be to have a specific `// expect: ...` comment that would do the opposite of `// ignore: ...`

Here is a typical example usage:

```dart
// expect: invalid_assignment
int number = 'Hello world';
```

The code would still not be executable. But the analysis report gets reversed:

`dartanalyzer` would report no issue, but only if the line followed by the `// expect:` has the mentioned issue.

**Comment by srawlins** [comment]
Can you give an example of "wanting to test that a specific code does not compile"? This does not sound like a typical use case, as a package author, and I cannot think of an example.

**Comment by rrousselGit** [comment]
Sure

One example would be for my package [Freezed](https://github.com/rrousselgit/freezed).\
It is a code-generator that wants to generate "compile-safe" code

It has multiple tests that assert that a given code compiles or not.

For example, one of the test I want to write for a new feature would be:

```dart
import 'package:freezed_annotation/freezed_annotation.dart';
part 'main.freezed.dart';

@freezed
abstract class CommonPropertyOfDifferentType with _$CommonPropertyOfDifferentType {
  factory CommonPropertyOfDifferentType.first(int value) = _First;
  factory CommonPropertyOfDifferentType.second(double value) = _First;
}

void main() {
  final obj = CommonPropertyOfDifferentType.first(42);
  // This should compile
  num sharedType = obj.value;
  // expect: invalid_assignment
  int firstType = obj.value;
  // expect: invalid_assignment
  double firstType = obj.value;
}
```
```

**Comment by rrousselGit** [comment]
I have another package in a similar situation that is not yet publicly available.
TL;DR it's a compile-safe service-locator.

So once again, I have to test the "compile-safe" aspect of the package.

One test for it would be:

```dart
void main() {
  final autoDispose = AutoDisposeProvider((_) => 42);

  // expect: invalid_assignment
  autoDispose.overrideForSubtree(
    Provider((_) => 42);
  );

  final normal = Provider((_) => 42);
  // the reverse scenario works
  normal.overrideForSubtree(
    AutoDisposeProvider((_) => 42);
  );
}
```

**Comment by srawlins** [comment]
I see, interesting. You could take a look at the analyzer's tests to see how we test analysis results. For example, take [test/src/diagnostics/assignment_to_const_test.dart](https://github.com/dart-lang/sdk/blob/master/pkg/analyzer/test/src/diagnostics/assignment_to_const_test.dart), and follow  `assertErrorsInCode` to `resolveTestFile` to [`DriverResolutionTest.resolveFile`](https://github.com/dart-lang/sdk/blob/master/pkg/analyzer/test/src/dart/resolution/driver_resolution.dart) to `AnalysisDriver.getResult`. You could copy some of these utilities, and use the analyzer's API to verify static errors.

Additionally you could look to see how the linter package tests lints with `// LINT` comments similar to your idea. [here](https://github.com/dart-lang/linter/blob/master/test/util/annotation_test.dart) are their unit tests and [here](https://github.com/dart-lang/linter/blob/master/test/rules/always_declare_return_types.dart) is an example of the linter tests.

**Comment by rrousselGit** [comment]
This is already what I do. The problem is the difficulty of that task.

The current solutions requires us to lose the help from all toolings and are brittle. 

A `// expect: error` is a lot cheaper to write, benefits from the tooling, and more resilient.
